base/array.jl

```diff
@@ -255,14 +255,14 @@ else
 end
 _default_eltype{I,T}(::Type{Generator{I,Type{T}}}) = T
 
-_array_for(T, itr, ::HasLength) = Array(T, Int(length(itr)::Integer))
-_array_for(T, itr, ::HasShape) = Array(T, convert(Dims,size(itr)))
+_array_for(T, itr, ::HasLength) = Array{T,ndims(itr)}(Int(length(itr)::Integer))
+_array_for(T, itr, ::HasShape) = Array{T,ndims(itr)}(convert(Dims,size(itr)))
 
```

Comment:
I had no idea that the `ndims(itr)` would ruin the type inference. @vtjnash Is there a way to specify the `N` in this case that preserves inferability?